addons/hr_evaluation/hr_evaluation.py

```diff
@@ -1,375 +1,465 @@
+        emp_obj = self.env['hr.employee']
+        for evl_rec in self:
+            if self.state == 'new' and vals.get('state') == 'done':
+                raise UserError(_("""Sorry ! You cannot drag this card from the "To Start" column to the "Done" column. You have to drag it to the "Appraisal Sent" column first."""))
+            # avoid recursive process
+            if vals.get('state') == 'pending' and not evl_rec._context.get('send_mail_status'):
+                evl_rec.button_send_appraisal()
+            if vals.get('interview_deadline') and self.state == 'pending' and not vals.get('meeting_id'):
+                if datetime.datetime.now().strftime(DF) > vals.get('interview_deadline'):
+                    raise UserError(_("The interview date can not be in the past"))
+                # creating employee meeting and interview date
+                evl_rec.create_update_meeting(vals)
+            if vals.get('appraisal_manager_ids'):
+                # add followers
+                user_ids = [employee.user_id.id for employee in emp_obj.browse(vals['appraisal_manager_ids'][0][2]) if employee.user_id]
+                evl_rec.message_subscribe_users(user_ids)
+        return super(hr_evaluation, self).write(vals)
+
+    @api.one
+    def update_appraisal_survey_url(self, survey_url, email_to, token):
+        survey_url = urlparse.urlparse(survey_url).path[1:]
+        if token:
+            survey_url = survey_url + '/' + token[0]
+        self.appraisal_url = survey_url
+        self.email_to = email_to
         return True
 
-    def survey_req_done(self, cr, uid, ids, context=None):
-        for id in self.browse(cr, uid, ids, context=context):
-            flag = False
-            wating_id = 0
-            if not id.evaluation_id.id:
-                raise UserError(_("You cannot start evaluation without Appraisal."))
-            records = id.evaluation_id.survey_request_ids
-            for child in records:
-                if child.state == "draft":
-                    wating_id = child.id
-                    continue
-                if child.state != "done":
-                    flag = True
-            if not flag and wating_id:
-                self.survey_req_waiting_answer(cr, uid, [wating_id], context=context)
-        self.write(cr, uid, ids, {'state': 'done'}, context=context)
+    @api.one
+    def create_token(self, email, survey, partner_id):
+        """ Create response with token """
+        token = uuid.uuid4().__str__()
+        self.env['survey.user_input'].create({
+            'survey_id': survey.id,
+            'deadline': self.date_close,
+            'date_create': datetime.datetime.now().strftime(DF),
+            'type': 'link',
+            'state': 'new',
+            'token': token,
+            'evaluation_id': self.id,
+            'partner_id': partner_id,
+            'email': email})
+        return token
+
+    @api.multi
+    def get_partner_ids(self, appraisal_receiver):
```

Comment:
This method name, behavior and variable names do not match and make it quite difficult to understand.

Should'nt you have a method 'send_appraisal' ?